src/main/java/org/checkerframework/specimin/unsolved/FullyQualifiedNameGenerator.java

```diff
@@ -376,21 +378,61 @@ else if (expr.isCastExpr()) {
 
     // Handle the cases where the type of the expression can be inferred from surrounding context
     if (expr.hasParentNode()) {
-      @Nullable FullyQualifiedNameSet fromLHS = getFQNsFromSurroundingContextType(expr);
+      @Nullable FullyQualifiedNameSet fromLHS = getFQNsFromSurroundingContextType(expr, canRecurse);
       if (fromLHS != null) {
         return fromLHS;
       }
     }
 
+    // Handle binary expressions after surrounding context because the binary expression could be on
+    // the
+    // right-hand side where the left side type is known.
+    if (expr.isBinaryExpr()) {
+      BinaryExpr binary = expr.asBinaryExpr();
+      Operator operator = binary.getOperator();
+
+      // Boolean
+      if (operator == Operator.AND
+          || operator == Operator.OR
+          || operator == Operator.EQUALS
+          || operator == Operator.NOT_EQUALS
+          || operator == Operator.LESS
+          || operator == Operator.GREATER
+          || operator == Operator.LESS_EQUALS
+          || operator == Operator.GREATER_EQUALS) {
+        return new FullyQualifiedNameSet("boolean");
+      } else {
+        // Treat all other cases; type on one side is equal to the other
+        FullyQualifiedNameSet leftType = getFQNsForExpressionTypeImpl(binary.getLeft(), canRecurse);
+        FullyQualifiedNameSet rightType =
+            getFQNsForExpressionTypeImpl(binary.getRight(), canRecurse);
+
+        // Remaining operators only work with primitive/String types
```

Comment:
What about boxed types like `java.lang.Integer`? Those are also possible here, although maybe you can just treat them as their primitive equivalents.